### PR TITLE
refactor theme edit feature

### PR DESCRIPTION
**Pull request by optizone**
This PR provides new theme editor(you can see full description of each option via tooltip): 
example

**Note**: You can edit only custom themes, so there are no radare2 standard themes in combo box at the top.

**Note №2**: Preview widget does not allow using context menu and some shortcuts so it can not change real disasm.

Current option is blinking in preview widget:
blink_example

If you changed some option, you can reset changes:
reset-example

Changes are instantly applied for preview theme:
pick-example

Implemented features:
- import/export theme
- rename theme 

Way of working with themes is changed, so all known (by me) bugs with copying, editing and displaying are fixed.

Fixed bugs:
- when copy a radare2 theme background turns black as well as a lot of other color options + highlight and highlightWord are copied wrong because they have alpha channel which is not supported by hex color representation
- user can edit radare2 standard color theme
- some bugs related to renaming (I cant remember what exactly xD)
- not all colors are displayed correctly (for instance highlight and highlightWord that have alpha channel)
- when copying theme Cutter does not make sure that all possible radare2 color options are provided by this theme

Other fixes:
- color options that Cutter does not use(cli graph related, ai.* options etc) are removed from edit dialog
- removed options highlight and highlightWord because they override radare2's linehl and wordhl and there is no point in it
- added some color option descriptions 

Appearance dialog is changed and now looks like this(there is a tooltip for each button):
appearence-dial

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
This is amazing. It looks very good and I 0layed with it recently. I am super happy about it.
In the next days I'll review this in code and strings level. Thank you!

**Comment by optizone** [comment]
This also needs to be checked on other platforms to make sure all widgets looks nice with different system color palettes.

File: src/common/ColorThemeWorker.cpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        .arg(r_sys_prefix(nullptr))
+        .arg(R_SYS_DIR)
+        .arg(R2_THEMES)
+    };
+    if (currDir.exists()) {
+        standardR2ThemesLocationPath = currDir.absolutePath();
+    } else {
+        QMessageBox::critical(nullptr,
+            tr("Standard themes not found"),
+            tr("The radare2 standard themes could not be found! "
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: "found!" -> "found."

File: src/common/ColorThemeWorker.cpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    return save(getTheme(srcThemeName), copyThemeName);
+}
+
+QString ColorThemeWorker::save(const QJsonDocument &theme, const QString &themeName) const
+{
+    QFile fOut(QDir(customR2ThemesLocationPath).filePath(themeName));
+    if (!fOut.open(QFile::WriteOnly | QFile::Truncate)) {
+        return tr("File <b>%1</b> can not be opened.")
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: "File" -> "The file"
"can not" -> "cannot"
is preferred

File: src/common/ColorThemeWorker.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+{
+    if (!isThemeExist(srcThemeName)) {
+        return tr("Theme <b>%1</b> does not exist.")
+                .arg(srcThemeName);
+    }
+
+    return save(getTheme(srcThemeName), copyThemeName);
+}
+
+QString ColorThemeWorker::save(const QJsonDocument &theme, const QString &themeName) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            fOut.write(line.arg(it.key())
+                       .arg(QColor(arr[0].toInt(), arr[1].toInt(), arr[2].toInt()).name()).toUtf8());
+        }
+    }
+
+    fOut.close();
+    return "";
+}
+
+bool ColorThemeWorker::isCustomTheme(const QString &themeName) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+{
+    return QFile::exists(QDir(customR2ThemesLocationPath).filePath(themeName));
+}
+
+bool ColorThemeWorker::isThemeExist(const QString &name) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+{
+    QStringList themes = Core()->getColorThemes();
+    return themes.contains(name);
+}
+
+QJsonDocument ColorThemeWorker::getTheme(const QString& themeName) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        }
+    }
+
+    for (auto &key : radare2UnusedOptions) {
+        theme.remove(key);
+    }
+    return QJsonDocument(theme);
+}
+
+QString ColorThemeWorker::deleteTheme(const QString &themeName) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                .arg(QFileInfo(file).filePath());
+    }
+    if (!file.remove()) {
+        return tr("File <b>%1</b> can not be removed.")
+                .arg(QFileInfo(file).filePath());
+    }
+    return "";
+}
+
+QString ColorThemeWorker::importTheme(const QString& file) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+         return "";
+     } else {
+         return tr("Error occurred during importing. "
+                   "Please make sure you have an access to "
+                   "the directory <b>%1</b> and try again.")
+                 .arg(src.dir().path());
+    }
+}
+
+QString ColorThemeWorker::renameTheme(const QString& themeName, const QString& newName) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+     QDir dir = customR2ThemesLocationPath;
+     bool ok = QFile::rename(dir.filePath(themeName), dir.filePath(newName));
+     if (!ok) {
+         return tr("Something went wrong during renaming. "
+                   "Please make sure you have access to the directory <b>\"%1\"</b>.").arg(dir.path());
+     }
+     return "";
+}
+
+bool ColorThemeWorker::isFileTheme(const QString& filePath, bool* ok) const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+            *ok = true;
+            return false;
+        }
+    }
+
+    *ok = true;
+    return true;
+}
+
+QStringList ColorThemeWorker::customThemes() const
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: documentation using docstrings

File: src/dialogs/preferences/ColorThemeEditDialog.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+#include "common/ColorThemeWorker.h"
+#include "common/Configuration.h"
+
+#include "widgets/ColorThemeListView.h"
+#include "widgets/DisassemblyWidget.h"
+
+#include <QScreen>
+#include <QKeyEvent>
+
+ColorThemeEditDialog::ColorThemeEditDialog(QWidget *parent) :
```
Comment (ITAYC0HEN) [inline src/dialogs/preferences/ColorThemeEditDialog.cpp]: documentation using docstrings

File: src/common/ColorThemeWorker.cpp [hunk 13]
```diff
@@ -0,0 +1,309 @@
+#include "ColorThemeWorker.h"
+
+#include <QRgb>
+#include <QDir>
+#include <QDebug>
```
Comment (ITAYC0HEN) [inline src/common/ColorThemeWorker.cpp]: is this necessary?

File: src/widgets/ColorPicker.cpp [hunk 14]
```diff
@@ -0,0 +1,334 @@
+#include "ColorPicker.h"
+#include "ui_ColorPicker.h"
+
+#include <QPaintEvent>
+#include <QPainter>
+#include <QtDebug>
```
Comment (ITAYC0HEN) [inline src/widgets/ColorPicker.cpp]: is this necessary?

File: src/widgets/ColorPicker.cpp [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+#include <QMouseEvent>
+#include <QDesktopWidget>
+#include <QPixmap>
+#include <QCursor>
+#include <QScreen>
+
+using namespace ColorPickerHelpers;
```
Comment (ITAYC0HEN) [inline src/widgets/ColorPicker.cpp]: Please document `ColorPicker.cpp`. Document the functions and add comments where needed in the code.
If this file was copied from other place, make sure the license is good for us

File: src/widgets/ColorThemeComboBox.cpp [hunk 16]
```diff
@@ -0,0 +1,85 @@
+#include "ColorThemeComboBox.h"
+
+#include "core/Cutter.h"
+
+#include "common/ColorThemeWorker.h"
+#include "common/Configuration.h"
+
+static const QHash<QString, ColorFlags> kRelevantThemes = {
```
Comment (ITAYC0HEN) [inline src/widgets/ColorThemeComboBox.cpp]: We already have such list in "AppearanceOptionsWidget". Try not duplicate and put it in mutual place

File: src/widgets/ColorThemeComboBox.cpp [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    { "solarized", DarkFlag },
+    { "zenburn", DarkFlag },
+    { "cutter", LightFlag },
+    { "dark", LightFlag },
+    { "matrix", LightFlag },
+    { "tango", LightFlag },
+    { "white", LightFlag }
+};
+
+ColorThemeComboBox::ColorThemeComboBox(QWidget *parent) : QComboBox(parent), showOnlyCustom(false)
```
Comment (ITAYC0HEN) [inline src/widgets/ColorThemeComboBox.cpp]: Please document this class and the functions inside

File: src/dialogs/preferences/ColorThemeEditDialog.cpp [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    QDialog::accept();
+}
+
+void ColorThemeEditDialog::reject()
+{
+    if (ThemeWorker().getTheme(Config()->getColorTheme()) !=
+        qobject_cast<ColorSettingsModel*>(ui->colorThemeListView->model())->getTheme() &&
+        QMessageBox::question(this,
+                              tr("Unsaved changes"),
+                              tr("Are you sure you want exit without saving? "
```
Comment (ITAYC0HEN) [inline src/dialogs/preferences/ColorThemeEditDialog.cpp]: "want exit" -> "want to exit"

**Comment by optizone** [comment]
@ITAYC0HEN do you mean you can see how context menu is being shown for a moment and then closes. Or that it doesn't even appear on the screen.
I want to remember that I disabled context menu in preview so user can not affect binary via it

**Comment by ITAYC0HEN** [comment]
Closes immediately. I can see it for brief a second

**Comment by optizone** [comment]
Every time you click on it?

**Comment by ITAYC0HEN** [comment]
Every time I right-clicked inside the preview, I can see the menu for a
brief of a moment and then closing

**Comment by xarkes** [comment]
I will review this tonight.

**Comment by ITAYC0HEN** [comment]
> I will review this tonight.

please make sure that @optizone is pushing his new changes and fixes before you review

**Comment by optizone** [comment]
@xarkes Im sick rn so wont be able to push in few days. I'll ping you when done

**Comment by xarkes** [comment]
Sure, rest well :)

**Comment by optizone** [comment]
ping @ITAYC0HEN @xarkes

**Comment by xarkes** [comment]
Could you rebase on master so we can merge? Thanks :)

File: src/Main.cpp [hunk 19]
```diff
@@ -48,6 +51,27 @@ int main(int argc, char *argv[])
 
     CutterApplication a(argc, argv);
 
+    // Removes obsolete color options (highlight and highlightWord) from custom theme files
+    if (!settings.value("updated_custom_themes", false).toBool()) {
+        const QStringList options = Core()->cmdj("ecj").object().keys()
+                                    << ColorThemeWorker::cutterSpecificOptions;
+        for (auto theme : Core()->cmd("eco*").split('\n', QString::SkipEmptyParts)) {
```
Comment (pelijah) [inline src/Main.cpp]: We have `cmdList()` for it

**Comment by ITAYC0HEN** [comment]
Thank you very much!! This is wonderful!

**Pull request merged**